### PR TITLE
Migrate PCA related tests from boost to catch2

### PR DESCRIPTION
**Pull request by mikkac**
Refers #2523

### CODE REVIEW

File: src/mlpack/tests/pca_test.cpp [hunk 1]
```diff
@@ -317,25 +314,22 @@ BOOST_AUTO_TEST_CASE(PCAScalingTest)
...
   // the first (plus 20% tolerance).
-  BOOST_REQUIRE_CLOSE(std::abs(eigvec(0, 0)), sqrt(2) / 2, 0.35);
-  BOOST_REQUIRE_CLOSE(std::abs(eigvec(1, 0)), sqrt(2) / 2, 0.35);
-  BOOST_REQUIRE_SMALL(eigvec(2, 0), 0.1); // Large tolerance for noise.
+  REQUIRE(std::abs(eigvec(0, 0)) == Approx(sqrt(2) / 2).epsilon(0.0035));
+  REQUIRE(std::abs(eigvec(1, 0)) == Approx(sqrt(2) / 2).epsilon(0.0035));
+  REQUIRE(eigvec(2, 0) == Approx(0.0).margin(0.1)); // Large tolerance for noise.
 
   // The eigenvalues should sum to three.
-  BOOST_REQUIRE_CLOSE(accu(eigval), 3.0, 0.1); // 10% tolerance.
```
Comment (mikkac) [inline src/mlpack/tests/pca_test.cpp:337]: `BOOST_REQUIRE_CLOSE(accu(eigval), 3.0, 0.1); // 10% tolerance.` 
Isn't `0.1` tolerance of level 0.1% here?
That's why in catch2's `REQUIRE` I set this to 0.1- here indeed it's 10%

File: src/mlpack/tests/pca_test.cpp [hunk 2]
```diff
@@ -317,25 +314,22 @@ BOOST_AUTO_TEST_CASE(PCAScalingTest)
...
-  BOOST_REQUIRE_SMALL(eigvec(0, 1), 0.1);
-  BOOST_REQUIRE_SMALL(eigvec(1, 1), 0.1);
-  BOOST_REQUIRE_CLOSE(std::abs(eigvec(2, 1)), 1.0, 0.35);
+  REQUIRE(eigvec(0, 1) == Approx(0.0).margin(0.1));
+  REQUIRE(eigvec(1, 1) == Approx(0.0).margin(0.1));
+  REQUIRE(std::abs(eigvec(2, 1)) == Approx(1.0).epsilon(0.0035));
 
   // The third component should have the same absolute value characteristics as
   // the first (plus 20% tolerance).
-  BOOST_REQUIRE_CLOSE(std::abs(eigvec(0, 0)), sqrt(2) / 2, 0.35);
```
Comment (mikkac) [inline src/mlpack/tests/pca_test.cpp:332]: Like in the comment below, here the tolerance is equal to 0.35% (according to instructions in #2523). What should we do here? Change comment or change code? Tests pass so I guess comment should be changed

Comment (rcurtin) [reply src/mlpack/tests/pca_test.cpp:332]: @mikkac I think the comment got out of date---it should say 0.35% like you pointed out.  My guess is that it started at 0.2, and whoever wrote the comment made an error thinking that represented 20%, then later the tolerance was increased to reduce random failures but the comment was forgotten. :smile: Thanks for noticing and pointing it out! :+1:

File: src/mlpack/tests/pca_test.cpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/tests/pca_test.cpp:337]: I think like the other one, the comment is wrong, so you can go with the original implementation's value and update the comment. :+1:

File: src/mlpack/tests/main_tests/pca_test.cpp [hunk 3]
```diff
@@ -96,23 +93,21 @@ BOOST_AUTO_TEST_CASE(PCANoVarRetainTest)
...
-  BOOST_REQUIRE_EQUAL(IO::GetParam<arma::mat>("output").n_cols, 5);
+  REQUIRE(IO::GetParam<arma::mat>("output").n_rows == 1);
+  REQUIRE(IO::GetParam<arma::mat>("output").n_cols == 5);
 }
 
 /**
  * Check that we can't specify an invalid new dimensionality.
  */
-BOOST_AUTO_TEST_CASE(PCATooHighNewDimensionalityTest)
+TEST_CASE_METHOD(PCATestFixture, "PCATooHighNewDimensionalityTest", "[PCAMainTest]")
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/pca_test.cpp:103]: Looks like this line is longer than 80 characters; would you mind wrapping it? :) I'm surprised the style checker didn't catch the issue.

There may be a few other lines where this is true---I didn't check them all (I'm using a phone right now so it's tedious :)).

File: src/mlpack/tests/main_tests/pca_test.cpp [hunk 4]
```diff
@@ -41,12 +40,10 @@ struct PCATestFixture
   }
 };
 
-BOOST_FIXTURE_TEST_SUITE(PCAMainTest, PCATestFixture);
-
 /**
  * Make sure that if we ask for a dataset in 3 dimensions back, we get it.
  */
-BOOST_AUTO_TEST_CASE(PCADimensionTest)
+TEST_CASE_METHOD(PCATestFixture, "PCADimensionTest", "[PCAMainTest]")
```
Comment (zoq) [inline src/mlpack/tests/main_tests/pca_test.cpp:46]: For the main tests we introduced another tag - `BindingTests` see: 

https://github.com/mlpack/mlpack/blob/6ced8d236590c86249d66204d8c1991beeb61249/src/mlpack/tests/main_tests/kmeans_test.cpp#L52-L53

for an example, can you add the tag.

File: src/mlpack/tests/main_tests/pca_test.cpp [hunk 5]
```diff
@@ -96,23 +93,22 @@ BOOST_AUTO_TEST_CASE(PCANoVarRetainTest)
...
-  BOOST_REQUIRE_EQUAL(IO::GetParam<arma::mat>("output").n_cols, 5);
+  REQUIRE(IO::GetParam<arma::mat>("output").n_rows == 1);
+  REQUIRE(IO::GetParam<arma::mat>("output").n_cols == 5);
 }
 
 /**
  * Check that we can't specify an invalid new dimensionality.
  */
-BOOST_AUTO_TEST_CASE(PCATooHighNewDimensionalityTest)
+TEST_CASE_METHOD(PCATestFixture, "PCATooHighNewDimensionalityTest",
```
Comment (zoq) [inline src/mlpack/tests/main_tests/pca_test.cpp:103]: I think there is no style guideline for this, but in the other tests the second line aligns with the first argument from the first one.

https://github.com/mlpack/mlpack/blob/6ced8d236590c86249d66204d8c1991beeb61249/src/mlpack/tests/main_tests/kmeans_test.cpp#L52-L53

is one example.

**Review by zoq — COMMENTED** [review]
Nice, just left a couple minor comments about the style.

**Comment by mikkac** [comment]
Any idea why those R-related checks fail? TBH I don't know whether I shoud take any additional steps to make it work

**Comment by rcurtin** [comment]
@mikkac I think these are related to #2602; can you merge master into your branch?  Once you do that, I think the builds should succeed. :+1:  (However it is not related to your code here I believe.)

**Review by birm — APPROVED** [review]
Passing tests!

**Pull request merged**